### PR TITLE
feat: help for core flags

### PR DESCRIPTION
**Pull request by rishabh3112**
**What kind of change does this PR introduce?**
Feature

**Did you add tests for your changes?**
No, will need

**If relevant, did you update the documentation?**
No

**Summary**
Adds help of flags as we have help for command from #1098.

**Does this PR introduce a breaking change?**
No

**Other information**
#1098 https://github.com/webpack/webpack-cli/issues/1053

### CODE REVIEW

File: lib/groups/help.js [hunk 1]
```diff
@@ -3,13 +3,13 @@ const chalk = require('chalk');
 const commandLineUsage = require('command-line-usage');
 
 class HelpGroup {
-    outputHelp(command = null) {
-        if (command !== null) {
-            const commands = require('../utils/commands').info;
+    outputHelp(subject = null, isCommand = true) {
+        if (subject !== null) {
+            const info = isCommand ? require('../utils/commands').info : require('../utils/core-flags').info;
+            const options = isCommand ? info.find(cmd => cmd.name === subject || cmd.alias === subject) : info.find(flg => flg.name === subject.slice(2) || flg.alias === subject.slice(1));
```
Comment (jamesgeorge007) [inline lib/groups/help.js]: This one seems a bit hard to read. Maybe use an if/else statement instead.

Comment (jamesgeorge007) [inline lib/groups/help.js]: This one seems a bit hard to read. Maybe use an if/else statement instead.

File: lib/utils/core-flags.js [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+        if (name === 'help') return [];
+        if (alias) {
+            return [`--${name}`, `-${alias}`];
+        }
+        return [`--${name}`];
+    })
+    .reduce((arr, val) => arr.concat(val), []);
+
+module.exports.names = names;
+module.exports.info = core;
```
Comment (jamesgeorge007) [inline lib/utils/core-flags.js]:
```js
module.exports = {names, core};
```

File: lib/webpack-cli.js [hunk 3]
```diff
@@ -162,11 +162,17 @@ class WebpackCLI extends GroupHelper {
...
-        return new HelpGroup().outputHelp(command);
+        const commandNames = require('./utils/commands').names;
+        const flagNames = require('./utils/core-flags').names;
+        const subject =
+            commandNames.filter(name => {
+                return args.includes(name);
+            })[0] ||
+            flagNames.filter(name => {
+                return args.includes(name);
+            })[0];
```
Comment (jamesgeorge007) [inline lib/webpack-cli.js]: Maybe there would be a potential scope for refactoring.

Comment (jamesgeorge007) [reply lib/webpack-cli.js]: Also, why not make use of the inline approach?

File: lib/groups/help.js (hunk 1 above)
Comment (rishabh3112) [reply lib/groups/help.js]: `if/else` would make `options` mutable. Instead used arrow function so as to not lose immutability.

File: lib/groups/help.js [hunk 4]
```diff
@@ -3,13 +3,18 @@ const chalk = require('chalk');
...
-        if (command !== null) {
-            const commands = require('../utils/commands').info;
+    outputHelp(subject = null, isCommand = true) {
+        if (subject !== null) {
+            const info = isCommand ? require('../utils/commands').info : require('../utils/core-flags').info;
+            const options = () => {
+                if (isCommand) {
+                    return info.find(cmd => cmd.name === subject || cmd.alias === subject);
+                }
+                return info.find(flg => flg.name === subject.slice(2) || flg.alias === subject.slice(1));
```
Comment (jamesgeorge007) [inline lib/groups/help.js]:
```suggestion
                return info.find(flag => flag.name === subject.slice(2) || flag.alias === subject.slice(1));
```

File: lib/groups/help.js (hunk 1 above)
Comment (jamesgeorge007) [reply lib/groups/help.js]: Yeah, makes sense :+1:

File: lib/groups/help.js (hunk 4 above)
Comment (rishabh3112) [reply lib/groups/help.js]: Made this change, also updated `cmd` to `command` too.

File: lib/webpack-cli.js (hunk 3 above)
Comment (rishabh3112) [reply lib/webpack-cli.js]: @jamesgeorge007 refactored in latest commit.

File: lib/utils/core-flags.js (hunk 2 above)
Comment (rishabh3112) [reply lib/utils/core-flags.js]: I prefer current ones as I want to keep `info` and variable name instead of `core`.

**Comment by rishabh3112** [comment]
@jamesgeorge007 awesome quick reviews 🐱‍👤👍

File: lib/utils/core-flags.js (hunk 2 above)
Comment (pranshuchittora) [reply lib/utils/core-flags.js]:
```

module.exports = {
    names,
    info:core
}

```

File: lib/groups/help.js [hunk 5]
```diff
@@ -3,13 +3,18 @@ const chalk = require('chalk');
 const commandLineUsage = require('command-line-usage');
 
 class HelpGroup {
-    outputHelp(command = null) {
-        if (command !== null) {
-            const commands = require('../utils/commands').info;
+    outputHelp(subject = null, isCommand = true) {
+        if (subject !== null) {
+            const info = isCommand ? require('../utils/commands').info : require('../utils/core-flags').info;
```
Comment (jamesgeorge007) [inline lib/groups/help.js]: https://github.com/webpack/webpack-cli/blob/3d6c24b8e345e7db48a5970d778f29168d5d3ba6/lib/utils/commands.js#L12

https://github.com/webpack/webpack-cli/blob/3d6c24b8e345e7db48a5970d778f29168d5d3ba6/lib/groups/help.js#L8

#1122

File: lib/webpack-cli.js (hunk 3 above)
Comment (jamesgeorge007) [reply lib/webpack-cli.js]: Much better :+1:

File: lib/utils/core-flags.js (hunk 2 above)
Comment (rishabh3112) [reply lib/utils/core-flags.js]: Cool. It's night here, will update tommorow 👍

Comment (jamesgeorge007) [reply lib/utils/core-flags.js]: We're all from the same time zone :sweat_smile:

File: lib/webpack-cli.js [hunk 6]
```diff
@@ -162,11 +162,14 @@ class WebpackCLI extends GroupHelper {
...
     runHelp(args) {
         const HelpGroup = require('./groups/help');
-        const { names } = require('./utils/commands');
-        const command = names.filter(name => {
+        const commandNames = require('./utils/commands').names;
+        const flagNames = require('./utils/core-flags').names;
+        const allNames = [...commandNames, ...flagNames];
+        const subject = allNames.filter(name => {
             return args.includes(name);
         })[0];
```
Comment (jamesgeorge007) [inline lib/webpack-cli.js]:
```js
        const subject = allNames.filter(name => args.includes(name))[0];
```
Maybe :thinking:

Comment (pranshuchittora) [reply lib/webpack-cli.js]: A little bit difficult to understand at first glance ...

**Comment by rishabh3112** [comment]
## ℹ Status 
- All CI errors related to this PR has been resolved.
- All expected features implemented

File: lib/utils/cli-flags.js [hunk 7]
```diff
@@ -182,27 +192,31 @@ module.exports = {
...
             name: 'hot',
+            usage: '--hot',
             alias: 'h',
             type: Boolean,
             group: ADVANCED_GROUP,
             description: 'Enables Hot Module Replacement',
         },
         {
             name: 'sourcemap',
+            usage: '--sourcemap <sourcemap>',
```
Comment (ematipico) [inline lib/utils/cli-flags.js]: Maybe we should show the alternatives provided by webpack:
`--sourcemap <sourcemap|eval|>` etc.

Basically what we have here: https://webpack.js.org/configuration/devtool/#devtool

File: lib/utils/cli-flags.js [hunk 8]
```diff
@@ -182,27 +192,31 @@ module.exports = {
         {
             name: 'target',
             alias: 't',
+            usage: '--target <build target>',
```
Comment (ematipico) [inline lib/utils/cli-flags.js]: Maybe we should show the targets available: https://webpack.js.org/configuration/target/#target

File: lib/utils/cli-flags.js [hunk 9]
```diff
@@ -129,50 +131,58 @@ module.exports = {
...
         {
             name: 'plugin',
+            usage: '--plugin',
             group: ADVANCED_GROUP,
             type: String,
             description: 'Load a given plugin',
         },
         {
             name: 'global',
+            usage: '--global',
```
Comment (ematipico) [inline lib/utils/cli-flags.js]: The usage should be different:
`--global myVar ./global.js`

File: lib/groups/help.js [hunk 10]
```diff
@@ -3,13 +3,18 @@ const chalk = require('chalk');
 const commandLineUsage = require('command-line-usage');
 
 class HelpGroup {
-    outputHelp(command = null) {
-        if (command !== null) {
-            const commands = require('../utils/commands').info;
+    outputHelp(subject = null, isCommand = true) {
```
Comment (ematipico) [inline lib/groups/help.js]: You should revert the properties. Usually the ones that are not needed (the ones that are `undefined` or `null`) should go after the one that have mandatory values.

Also, are you use you want to have a `null`? I mean, if you don't pass anything `isCommand` will be `undefined` which is good for this case, right?

File: lib/groups/help.js [hunk 11]
```diff
@@ -3,13 +3,18 @@ const chalk = require('chalk');
...
+        if (subject !== null) {
+            const info = isCommand ? require('../utils/commands').info : require('../utils/core-flags').info;
+            const options = (() => {
```
Comment (ematipico) [inline lib/groups/help.js]: Why this is and inline IIFE? I don't think we need it.

```js
const options = info.find(commandOrFlag => {
   if (isCommand) {
     // return 
   }
   // return
})
```

File: lib/utils/core-flags.js [hunk 12]
```diff
@@ -0,0 +1,13 @@
+const { core } = require('./cli-flags');
+
+const names = core
```
[the remark this answers is not in the document]
Comment (ematipico) [reply lib/utils/core-flags.js]: Maybe also a test would help. It's up to you! We should improve the readability/usage of the code :)

File: lib/groups/help.js (hunk 11 above)
Comment (rishabh3112) [reply lib/groups/help.js]: > Why this is

`options` is the `object` with details about `subject`.

> inline IIFE?

I was recommended to maintain immutability at each step and I didn't had any other implementation in my mind. Thus Opted for inline iffe. Will update it as per your suggestion in next commit.

> I don't think we need it.

Is there any other way to do this?

**Comment by rishabh3112** [comment]
I updated the help and added documentation links for some of those.
image

File: lib/groups/help.js (hunk 11 above)
Comment (ematipico) [reply lib/groups/help.js]: You can use `.reduce` if you really want to keep a functional approach :)

**Comment by ematipico** [comment]
@rishabh3112 if you resolved the conflicts, I'd say we're ready to merge it!

**Comment by rishabh3112** [comment]
@ematipico rebased, and we are ready to go I think.

**Pull request merged**